File: tasks/fastatofastq/src/convert.py
#!/usr/bin/env python

from __future__ import print_function

import sys
import os
import gzip

COLORMAP = {
    'A': [ 'A', 'C', 'G', 'T' ],
    'C': [ 'C', 'A', 'T', 'G' ],
    'G': [ 'G', 'T', 'A', 'C' ],
    'T': [ 'T', 'G', 'C', 'A' ]
}

def decode(seq):
    if len(seq) == 0: return ""
    last = seq[0]; retval = ""
    if last not in COLORMAP: return ""
    for i in range(1, len(seq)):
        try:
            last = COLORMAP[last][int(seq[i])] if last != 'N' else 'N'
        except:
            last = 'N'
        retval += last
    return retval

def qual(seq):
    retval = ""
    for x in seq.strip().split():
        retval += chr(int(x) + 32)
    return retval

def writeseq(name, seq, qual, o):
    o.write("@%s\n%s\n+\n%s\n" % (name.strip(), seq, qual))

def main(argc, argv):

    if argc < 4:
        print("usage: convert.py input.fasta input.csqual output.fastq", file = sys.stderr)
        return 1
    fastain = argv[1]
    qualin = argv[2]
    seqs = {}; quals = {}
    
    with (gzip.open if fastain.endswith(".gz") else open)(fastain, 'r') as fastafile:
        with (gzip.open if qualin.endswith(".gz") else open)(qualin, 'r') as qualfile:
            with gzip.open(argv[3] + (".gz" if not argv[3].endswith(".gz") else ""), 'w') as o:
                fname = fastafile.readline(); qname = qualfile.readline(); i = 0
                while fname and fname.startswith('#'):
                    fname = fastafile.readline()
                while qname and qname.startswith('#'):
                    qname = qualfile.readline()
                while fname and qname:
                    i += 1
                    if i % 100000 == 0: print(i)
                    if qname == fname:
                        writeseq(fname, decode(fastafile.readline()), qual(qualfile.readline()), o)
                    else:
                        if qname in seqs:
                            writeseq(qname, seqs[qname], qual(qualfile.readline()), o)
                            del seqs[qname]
                        else:
                            quals[qname] = qual(qualfile.readline())
                        if fname in quals:
                            writeseq(fname, decode(fastafile.readline()), quals[fname], o)
                            del quals[fname]
                        else:
                            seqs[fname] = decode(fastafile.readline())
                    qname = qualfile.readline()
                    fname = fastafile.readline()

    return 0

if __name__ == "__main__":
    sys.exit(main(len(sys.argv), sys.argv))

File: tasks/fastatofastq/src/convert.cpp
/**
 * convert.cpp
 * Converts a CSFASTA and CSQUAL pair to FASTQ format.
 */

#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <unordered_map>
#include <vector>

const std::unordered_map<char, std::vector<char>> COLORMAP = {
  { 'A', { 'A', 'C', 'G', 'T' } },
  { 'C', { 'C', 'A', 'T', 'G' } },
  { 'G', { 'G', 'T', 'A', 'C' } },
  { 'T', { 'T', 'G', 'C', 'A' } }
};

std::string decode(const std::string& fastaline) {
  char last = fastaline[0];
  if (COLORMAP.find(last) == COLORMAP.end()) last = 'N';
  std::string out = "";
  for (int i = 1; i < fastaline.size(); ++i) {
    last = fastaline[i] == '.' || last == 'N' ? 'N' : COLORMAP.find(last)->second[fastaline[i] - 48];
    out += last;
  }
  return out;
}

std::string dqual(const std::string& qualline) {
  std::string cint, out = "";
  std::stringstream s(qualline);
  while (std::getline(s, cint, ' ')) {
    out += (char)(cint.compare("-1") == 0 ? 33 : std::atoi(cint.c_str()) + 33);
  }
  return out;
}

int main(int argc, char **argv) {

  if (argc < 4) {
    std::cerr << "usage: fastatofastq input.csfasta input.csqual output.fastq\n";
    return 1;
  }

  size_t i = 0;
  std::ifstream fasta(argv[1]), qual(argv[2]);
  std::string fastaline, qualline, fastaseq, qualseq;
  std::ofstream output(argv[3]);
  std::unordered_map<std::string, std::string> fastas, quals;

  while (std::getline(fasta, fastaline) && std::getline(qual, qualline) && ++i) {
    
    if (i % 100000 == 0) std::cerr << i << '\n';
    
    // skip comments, break on empty line
    while (fastaline[0] == '#' && std::getline(fasta, fastaline));
    while (qualline[0] == '#' && std::getline(qual, qualline));
    if (fastaline.length() == 0 || qualline.length() == 0) break;

    // read actual data
    std::getline(fasta, fastaseq);
    std::getline(qual, qualseq);
    
    // if they are the same, just write to output
    if (fastaline.compare(qualline) == 0) {
      output << '@' << fastaline << '\n' << decode(fastaseq) << "\n+\n" << dqual(qualseq) << '\n';
    } else {

      // if this line from the csqual has already been found in the csfasta, write the pair;
      // otherwise, store it in the csqual map for later
      if (fastas.find(qualline) != fastas.end()) {
	output << '@' << qualline << '\n' << fastas[qualline] << "\n+\n" << dqual(qualseq) << '\n';
	fastas.erase(qualline);
      } else
	quals[qualline] = dqual(qualseq);

      // if this line from the csfasta has already been found in the csqual, write the pair;
      // otherwise, store it in the csfasta map for later
      if (quals.find(fastaline) != quals.end()) {
	output << '@' << fastaline << '\n' << decode(fastaseq) << "\n+\n" << quals[fastaline] << '\n';
	quals.erase(fastaline);
      } else
	fastas[fastaline] = decode(fastaseq);
      
    }
    
  }

  std::cerr << fastaline << '\n' << qualline << '\n';

  return 0;
  
}
